proposals/POEM-2-function-pulling-container-scheduler.md

```diff
@@ -0,0 +1,132 @@
+Previously we revealed [performance issues](https://cwiki.apache.org/confluence/display/OPENWHISK/Autonomous+Container+Scheduling+v1) in OpenWhisk.
+There are many reasons for them, but we can summarize them into the following problems.
+
+1. Multiple actions share the same `homeInvoker`.
+2. Current scheduler does not consider time taking for container operations(create/delete, pause/unpause).
+3. Resources are evenly divided by the number of controllers.
+
+First, multiple actions share the same `homeInvoker`. This is because the `homeInvoker` is statically decided by a hash function.
+While scheduling, resource status in the invoker side is not considered and this introduces busy hotspot invokers while the others are idle.
+This is generally a good idea to increase the probability of container reuse, but it becomes bad in the real scene because of slow container operation.
+When all resources in an invoker are taken by one action(busy/warm), an activation for another action will remove one of the existing containers and create a new one for it.
+If actions run for a long time this would be a good approach in terms of bin-packing. But when actions run for short time(most of the serverless cases),
+it severely increases the response time because container operations take one to two orders of magnitude longer time than invocation and it results in bad performance.
+
+Second, since controllers schedule activations to invokers in a distributed environment, controllers are assigned resource fraction of invokers.
+It's not feasible to collect the status of invokers, scheduling decisions made by all controllers and schedule based on them in real time so that controllers schedule activations within their own partial invoker resources.
+As a result, controllers throttle activations even if cluster-wide invocations are less than limits.
+
+We propose FPCScheduler to address the above issues.
+The scheduler differs from the existing scheduler(`ShardingPoolBalancer`) in many directions.
+It schedules containers rather than function requests. Each container is supposed to pull activations requests from the given action queue continuously.
+Whenever one execution is over, it fetches the next activation requests and invokes it repeatedly. In this way, we can maximize container reuse.
+One more benefit is schedulers don't need to track/consider the location of existing containers. Schedulers just decide where and when to create more containers.
+It can just create more containers in invokers with enough resources. And this enables distributed scheduling with all invoker resources among multiple schedulers.
+
+Controllers no longer schedule activation requests. They are in charge of creating action queues and forwarding activation requests to the queues.
+The action queue is a dedicated queue for the given action and dynamically created/deleted in schedulers. Each action has its own queue, so there is no interference among actions because of
```

Comment:
```suggestion
Each action queue is a dedicated queue for the given action and dynamically created/deleted by the FPCSchedulers. Each action has its own queue, so there is no interference among actions because of a shared queue.
```

Reply:
Is the queue specific to a function name or name + namespace?

Reply:
The queue is uniquely identified by the FQN of the given action.
So `namespace + (package name)+ name`.